docs/articles/building_application.md

```diff
@@ -251,8 +242,37 @@ sender.Send(
     (m, o, s) => Thread.Sleep(120000),
     sender);
 ```
-Thread.Sleep is a hypothetical example of having blocking calls in the MessageCallback.
- 
+
+The second occurs when an async operation is completed and its continuation runs *synchronously*:
+
+```
+SenderLink sender = new SenderLink(session, "sender", "q1");
+await sender.SendAsync( new Message("test"));
+
+Thead.Sleep(120000);
+```
+
+The above code will block the pump if the continuation of SendAsync happens to be run synchronously (which it frequently will be). 
+The problem is the same as with the first example, the async pump does not get a chance to release the currently executing thread back to the 
+thread pool and therefore has no chance to await further I/O. This means that no messages can be processed and will typically cause deadlock, 
+application hang or timeout errors.  
+
+Specifically it is important to understand that blocking operations to be avoided include the sync API of this library:
+```
+SenderLink sender = new SenderLink(session, "sender", "q1");
+await sender.SendAsync(new Message("m1"));
+sender.Send(new Message("m2"));
+```
+Here, the Send call will timeout because the returned acknowledgement cannot be processed when the I/O
+processing is blocked.
+
+The solution for callbacks is naturally to schedule blocking work asynchronously instead of blocking the calling thread. The solution for continuations 
+depends on the application, which must either ensure that continuations of async operations never occur on threads that do blocking work, or the async operations can be wrapped by the application using a `TaskCompletionSource` with `TaskCreationOptions.RunContinuationsAsynchronously` specified.
```

Comment:
Ideally an example on how to do this should be included, but I didn't want to add one in a rush in case I get it wrong :)